File: recursion/backtracking/n_queen.c
/*
 * 回溯算法 - N皇后问题
 * 在n×n棋盘上放置n个皇后，使其互不攻击
 */
#include <stdio.h>
#include <stdlib.h>

#define N 4  // 棋盘大小

// 打印棋盘
void printBoard(int board[N][N]) {
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            printf("%c ", board[i][j] ? 'Q' : '.');
        }
        printf("\n");
    }
    printf("\n");
}

// 检查在(row, col)放置皇后是否安全
int isSafe(int board[N][N], int row, int col) {
    // 检查列
    for (int i = 0; i < row; i++) {
        if (board[i][col]) return 0;
    }
    
    // 检查左上对角线
    for (int i = row - 1, j = col - 1; i >= 0 && j >= 0; i--, j--) {
        if (board[i][j]) return 0;
    }
    
    // 检查右上对角线
    for (int i = row - 1, j = col + 1; i >= 0 && j < N; i--, j++) {
        if (board[i][j]) return 0;
    }
    
    return 1;
}

// 使用回溯求解N皇后
int solveNQueen(int board[N][N], int row) {
    // 基础情况：所有皇后都已放置
    if (row >= N) {
        return 1;
    }
    
    // 尝试在当前行的每一列放置皇后
    for (int col = 0; col < N; col++) {
        if (isSafe(board, row, col)) {
            // 放置皇后
            board[row][col] = 1;
            
            // 递归：放置下一行的皇后
            if (solveNQueen(board, row + 1)) {
                return 1;
            }
            
            // 回溯：撤销放置
            board[row][col] = 0;
        }
    }
    
    return 0;
}

int main() {
    int board[N][N] = {0};
    
    printf("%d皇后问题求解:\n\n", N);
    
    if (solveNQueen(board, 0)) {
        printf("找到解:\n");
        printBoard(board);
    } else {
        printf("无解\n");
    }
    
    return 0;
}

File: recursion/backtracking/n_queen.py
"""
回溯算法 - N皇后问题
在n×n棋盘上放置n个皇后，使其互不攻击
"""

N = 4  # 棋盘大小

def print_board(board):
    """打印棋盘"""
    for i in range(N):
        row = ""
        for j in range(N):
            row += "Q " if board[i][j] == 1 else ". "
        print(row)
    print()

def is_safe(board, row, col):
    """检查在(row, col)放置皇后是否安全"""
    # 检查列
    for i in range(row):
        if board[i][col] == 1:
            return False
    
    # 检查左上对角线
    for i, j in range(row-1, col-1, -1), range(row-1, col-1, -1):
        if i >= 0 and j >= 0 and board[i][j] == 1:
            return False
    
    # 检查右上对角线
    for i, j in zip(range(row-1, -1, -1), range(col+1, N)):
        if board[i][j] == 1:
            return False
    
    return True

def solve_n_queen(board, row):
    """使用回溯求解N皇后"""
    # 基础情况：所有皇后都已放置
    if row >= N:
        return True
    
    # 尝试在当前行的每一列放置皇后
    for col in range(N):
        if is_safe(board, row, col):
            # 放置皇后
            board[row][col] = 1
            
            # 递归：放置下一行的皇后
            if solve_n_queen(board, row + 1):
                return True
            
            # 回溯：撤销放置
            board[row][col] = 0
    
    return False

def main():
    board = [[0 for _ in range(N)] for _ in range(N)]
    
    print(f"{N}皇后问题求解:\n")
    
    if solve_n_queen(board, 0):
        print("找到解:")
        print_board(board)
    else:
        print("无解")

if __name__ == "__main__":
    main()
